### PR TITLE
8254562: ZGC: Remove ZMarkRootsTask

### PR DESCRIPTION
**Pull request by stefank**
After introducing concurrent stack scanning, we don't need to mark through any roots during the mark start pause. Remove the code.

Note that the old code passed in `false` to `_roots(false /* visit_jvmti_weak_export */)`. This has the effect that no roots are visited by the iterator:
```
void ZRootsIterator::oops_do(ZRootsIteratorClosure* cl) {
  ZStatTimer timer(ZSubPhasePauseRoots);
  if (_visit_jvmti_weak_export) {
    _jvmti_weak_export.oops_do(cl);
  }
}
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8254562](https://bugs.openjdk.java.net/browse/JDK-8254562): ZGC: Remove ZMarkRootsTask

### Reviewers
 * [Per Liden](https://openjdk.java.net/census#pliden) (@pliden - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/601/head:pull/601`
`$ git checkout pull/601`

### CODE REVIEW

**Comment by stefank** [comment]
Looking at this some more I figured that we could clean up the rest of the ZRootIterator usages. The JVMTI cleaning isn't really a root, but rather a way for us to update the JVMTI datastructure after the objects have moved (update their identities).

So, the last patch:
1) Removes the unnecessary spawning of parallel threads in ZRelocate::start. The JVMTI roots are only visited in a single-threaded operation.
2) Removed/replaced ZRootsIterator. Neither ZVerify, nor ZHeapIterator, really used it, because they implicitly sent in `false` to the constructor (see above comment). ZRelocate now performs a direct calls, which makes it easier to understand, IMHO.

**Comment by pliden** [comment]
@stefank I see we're poking around in the same code :) I agree about the ZMarkTask/ZVerify/ZHeapIterator changes, but I don't think your change to ZRelocate will actually work, because without a ZTask the VM thread will not have a worker id, which will make the allocation path sad. To avoid stalling this patch, would you mind not doing the ZRelocate changes here, and we can continue discussning how we want to do this.

**Review by fisk — COMMENTED** [review]
I agree with Per that we can start by removing the marking stuff first. I think Coleen is just about to move the JVMTI tag map oops into OopStorage, and then we can nuke the relocation code too.

**Comment by stefank** [comment]
Talked to Per. The plan is to revert the removal of the ZRelocateTask, but at least run that code with only one worker thread. Running tests now.

File: src/hotspot/share/gc/z/zRelocate.cpp [hunk 1]
```diff
@@ -56,24 +56,26 @@ class ZRelocateRootsIteratorClosure : public ZRootsIteratorClosure {
...
 
   virtual void work() {
+    // Allocation path assumes that relocating GC threads are ZWorkers
+    assert(ZThread::is_worker(), "Relocation code needs to be run as a worker");
+    assert(ZThread::worker_id() == 0, "No multi-thread support");
+
     // During relocation we need to visit the JVMTI
     // export weak roots to rehash the JVMTI tag map
-    _roots.oops_do(&_cl);
+    ZRelocateRoots::oops_do(&_cl);
```
Comment (pliden) [inline src/hotspot/share/gc/z/zRelocate.cpp:72]: I would have been more explicit here, something like this:

```
class ZRelocateJVMTIWeakExportTask : public ZTask {
public:
  ZRelocateJVMTIWeakExportTask() :
      ZTask("ZRelocateJVMTIWeakExportTask") {}

  virtual void work() {
    AlwaysTrueClosure always_alive;
    ZRelocateRootsIteratorClosure _cl;
    JvmtiExport::weak_oops_do(&always_alive, &cl);
  }
};

void ZRelocate::start() {
  // During relocation we need to visit the JVMTI
  // export weak roots to rehash the JVMTI tag map
  ZStatTimer timer(ZSubPhasePauseRelocateJVMTIWeakExport);
  ZRelocateJVMTIWeakExportTask task;
  _workers->run_serial(&task);
}
```

But this code will go away soon anyway, so you decide.

**Review by pliden — APPROVED** [review]

File: src/hotspot/share/gc/z/zRelocate.cpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/gc/z/zRelocate.cpp:72]: I agree, and I actually started that way, but hit some minor snags that ZSubPhasePauseRelocateJVMTIWeakExport wasn't available to zRelocate.cpp, and I wasn't sure about moving a *PhasePause* instance out from the zRootsIterator.cpp.

**Comment by stefank** [comment]
I'm going to push the patch as is. The relocate part will change soon anyway.

/integrate

**Pull request closed without merging**